# v1 / circuit_d3: optO2 -> isa
# circuit_d3.4_optO2.ll (13747 lines) -> circuit_d3.5_isa.s (17802 lines)
# 2 largest change hunks, 6 lines context, capped at 80 lines/side. Full diff: diff -u circuit_d3.4_optO2.ll circuit_d3.5_isa.s

@@ hunk 1/2 left 1..10 right 1..80 @@
--- circuit_d3.4_optO2.ll:1
+++ circuit_d3.5_isa.s:1
@@ -1,10 +1,80 @@
-; ModuleID = 'V2_performance/lowering/v1/circuit_d3.3_translate.ll'
-source_filename = "LLVMDialectModule"
-target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-p9:192:256:256:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8:9"
-target triple = "amdgcn-amd-amdhsa"
-
-@lds_red_passed = external local_unnamed_addr addrspace(3) global [4 x i64]
-@lds_red_logical = external local_unnamed_addr addrspace(3) global [4 x i64]
-@lds_red_obs = external local_unnamed_addr addrspace(3) global [32 x i64]
-
-; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none)
+	.amdgcn_target "amdgcn-amd-amdhsa--gfx950"
+	.amdhsa_code_object_version 6
+	.text
+	.globl	clifft_log                      ; -- Begin function clifft_log
+	.p2align	2
+	.type	clifft_log,@function
+clifft_log:                             ; @clifft_log
+; %bb.0:
+	s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+	s_movk_i32 s0, 0xfc01
+	v_lshrrev_b32_e32 v2, 20, v1
+	v_mov_b32_e32 v3, 0
+	s_mov_b32 s1, -1
+	v_lshl_add_u64 v[2:3], v[2:3], 0, s[0:1]
+	v_cvt_f64_i32_e32 v[4:5], v3
+	v_and_b32_e32 v1, 0x800fffff, v1
+	v_ldexp_f64 v[4:5], v[4:5], 32
+	v_cvt_f64_u32_e32 v[2:3], v2
+	v_or_b32_e32 v1, 0x3ff00000, v1
+	v_add_f64 v[2:3], v[4:5], v[2:3]
+	v_add_f64 v[4:5], v[0:1], -1.0
+	v_add_f64 v[0:1], v[0:1], 1.0
+	v_div_scale_f64 v[6:7], s[0:1], v[0:1], v[0:1], v[4:5]
+	v_rcp_f64_e32 v[8:9], v[6:7]
+	s_mov_b32 s0, 0xfefa39ef
+	s_mov_b32 s1, 0x3fe62e42
+	v_mul_f64 v[2:3], v[2:3], s[0:1]
+	v_fma_f64 v[10:11], -v[6:7], v[8:9], 1.0
+	v_fmac_f64_e32 v[8:9], v[8:9], v[10:11]
+	v_fma_f64 v[10:11], -v[6:7], v[8:9], 1.0
+	v_fmac_f64_e32 v[8:9], v[8:9], v[10:11]
+	v_div_scale_f64 v[10:11], vcc, v[4:5], v[0:1], v[4:5]
+	v_mul_f64 v[12:13], v[10:11], v[8:9]
+	v_fma_f64 v[6:7], -v[6:7], v[12:13], v[10:11]
+	s_mov_b32 s0, 0x611a7b96
+	s_nop 0
+	v_div_fmas_f64 v[6:7], v[6:7], v[8:9], v[12:13]
+	v_div_fixup_f64 v[0:1], v[6:7], v[0:1], v[4:5]
+	v_mul_f64 v[4:5], v[0:1], v[0:1]
+	s_mov_b32 s1, 0x3fa1a7b9
+	v_mul_f64 v[6:7], v[4:5], s[0:1]
+	s_mov_b32 s0, 0xbda12f68
+	s_mov_b32 s1, 0x3fa2f684
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x47ae147b
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fa47ae1
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x590b2164
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fa642c8
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x18618618
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fa86186
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0xbca1af28
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3faaf286
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x1e1e1e1e
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fae1e1e
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x11111111
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fb11111
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x13b13b14
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fb3b13b
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x745d1746
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fb745d1
+	v_add_f64 v[6:7], v[6:7], s[0:1]
+	s_mov_b32 s0, 0x1c71c71c
+	v_mul_f64 v[6:7], v[4:5], v[6:7]
+	s_mov_b32 s1, 0x3fbc71c7
+	v_add_f64 v[6:7], v[6:7], s[0:1]

@@ hunk 2/2 left 1..80 right 17796..17802 @@
--- circuit_d3.4_optO2.ll:1
+++ circuit_d3.5_isa.s:17796
@@ -1,80 +1,7 @@
-; ModuleID = 'V2_performance/lowering/v1/circuit_d3.3_translate.ll'
-source_filename = "LLVMDialectModule"
-target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-p9:192:256:256:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8:9"
-target triple = "amdgcn-amd-amdhsa"
+amdhsa.target:   amdgcn-amd-amdhsa--gfx950
+amdhsa.version:
+  - 1
+  - 2
+...
 
-@lds_red_passed = external local_unnamed_addr addrspace(3) global [4 x i64]
-@lds_red_logical = external local_unnamed_addr addrspace(3) global [4 x i64]
-@lds_red_obs = external local_unnamed_addr addrspace(3) global [32 x i64]
-
-; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none)
-declare noundef range(i32 0, 1024) i32 @llvm.amdgcn.workitem.id.x() #0
-
-; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none)
-declare noundef i32 @llvm.amdgcn.workgroup.id.x() #0
-
-; Function Attrs: convergent mustprogress nocallback nofree nounwind willreturn
-declare void @llvm.amdgcn.s.barrier() #1
-
-; Function Attrs: convergent mustprogress nocallback nofree nounwind willreturn memory(none)
-declare i32 @llvm.amdgcn.ds.bpermute(i32, i32) #2
-
-; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
-define double @clifft_log(double %0) local_unnamed_addr #3 {
-  %2 = bitcast double %0 to i64
-  %3 = lshr i64 %2, 52
-  %4 = add nsw i64 %3, -1023
-  %5 = sitofp i64 %4 to double
-  %6 = fmul double %5, 0x3FE62E42FEFA39EF
-  %7 = and i64 %2, -9218868437227405313
-  %8 = or disjoint i64 %7, 4607182418800017408
-  %9 = bitcast i64 %8 to double
-  %10 = fadd double %9, -1.000000e+00
-  %11 = fadd double %9, 1.000000e+00
-  %12 = fdiv double %10, %11
-  %13 = fmul double %12, %12
-  %14 = fmul double %13, 0x3FA1A7B9611A7B96
-  %15 = fadd double %14, 0x3FA2F684BDA12F68
-  %16 = fmul double %13, %15
-  %17 = fadd double %16, 4.000000e-02
-  %18 = fmul double %13, %17
-  %19 = fadd double %18, 0x3FA642C8590B2164
-  %20 = fmul double %13, %19
-  %21 = fadd double %20, 0x3FA8618618618618
-  %22 = fmul double %13, %21
-  %23 = fadd double %22, 0x3FAAF286BCA1AF28
-  %24 = fmul double %13, %23
-  %25 = fadd double %24, 0x3FAE1E1E1E1E1E1E
-  %26 = fmul double %13, %25
-  %27 = fadd double %26, 0x3FB1111111111111
-  %28 = fmul double %13, %27
-  %29 = fadd double %28, 0x3FB3B13B13B13B14
-  %30 = fmul double %13, %29
-  %31 = fadd double %30, 0x3FB745D1745D1746
-  %32 = fmul double %13, %31
-  %33 = fadd double %32, 0x3FBC71C71C71C71C
-  %34 = fmul double %13, %33
-  %35 = fadd double %34, 0x3FC2492492492492
-  %36 = fmul double %13, %35
-  %37 = fadd double %36, 2.000000e-01
-  %38 = fmul double %13, %37
-  %39 = fadd double %38, 0x3FD5555555555555
-  %40 = fmul double %13, %39
-  %41 = fadd double %40, 1.000000e+00
-  %42 = fmul double %12, 2.000000e+00
-  %43 = fmul double %42, %41
-  %44 = fadd double %6, %43
-  ret double %44
-}
-
-; Function Attrs: nofree norecurse nosync nounwind memory(argmem: readwrite)
-define void @clifft_draw_next_noise(ptr captures(none) %0, ptr captures(none) %1, ptr readonly captures(none) %2, i64 %3) local_unnamed_addr #4 {
-  %5 = load i32, ptr %0, align 4
-  %6 = zext i32 %5 to i64
-  %.not = icmp ugt i64 %3, %6
-  br i1 %.not, label %7, label %96
-
-7:                                                ; preds = %4
-  %8 = icmp eq i32 %5, 0
-  br i1 %8, label %13, label %9
-
+	.end_amdgpu_metadata
